src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -494,67 +494,69 @@ private async Task ReleaseLeasesAsync(SqlConnection connection, CancellationToke
+                            }
 
-                        transaction.Commit();
+                            // Update the global state table if we have processed all changes with ChangeVersion <= newLastSyncVersion,
+                            // and clean up the leases table to remove all rows with ChangeVersion <= newLastSyncVersion.
+                            using (SqlCommand updateTablesPostInvocationCommand = this.BuildUpdateTablesPostInvocation(connection, transaction, newLastSyncVersion))
+                            {
+                                this._logger.LogDebugWithThreadId($"BEGIN UpdateTablesPostInvocation Query={updateTablesPostInvocationCommand.CommandText}");
+                                var commandSw = Stopwatch.StartNew();
+                                await updateTablesPostInvocationCommand.ExecuteNonQueryAsync(token);
+                                updateLastSyncVersionDurationMs = commandSw.ElapsedMilliseconds;
+                                this._logger.LogDebugWithThreadId($"END UpdateTablesPostInvocation Duration={updateLastSyncVersionDurationMs}ms");
+                            }
 
-                        var measures = new Dictionary<TelemetryMeasureName, double>
-                        {
-                            [TelemetryMeasureName.ReleaseLeasesDurationMs] = releaseLeasesDurationMs,
-                            [TelemetryMeasureName.UpdateLastSyncVersionDurationMs] = updateLastSyncVersionDurationMs,
-                        };
+                            transaction.Commit();
 
-                        TelemetryInstance.TrackEvent(TelemetryEventName.ReleaseLeasesEnd, this._telemetryProps, measures);
-                    }
-                    catch (Exception ex)
-                    {
-                        this._logger.LogError($"Failed to execute SQL commands to release leases for table '{this._userTable.FullName}' due to exception: {ex.GetType()}. Exception message: {ex.Message}");
-                        TelemetryInstance.TrackException(TelemetryErrorName.ReleaseLeases, ex, this._telemetryProps);
+                            var measures = new Dictionary<TelemetryMeasureName, double>
+                            {
+                                [TelemetryMeasureName.ReleaseLeasesDurationMs] = releaseLeasesDurationMs,
+                                [TelemetryMeasureName.UpdateLastSyncVersionDurationMs] = updateLastSyncVersionDurationMs,
+                            };
 
-                        try
-                        {
-                            transaction.Rollback();
+                            TelemetryInstance.TrackEvent(TelemetryEventName.ReleaseLeasesEnd, this._telemetryProps, measures);
+                            // Don't want to loop if no exception occurs.
+                            break;
                         }
-                        catch (Exception ex2)
+                        catch (Exception ex) when (retryCount < MaxRetryReleaseLeases)
```

Comment:
IMO, we should log the error and attempt to rollback the transaction event if this is not the last retry. The for-block can be placed outside the outer-most try-block.

Reply:
We should definitely be trying to rollback on the last exception. The error and exception event being logged though I think I would prefer to keep separate as they are - we want to make it clear that it failed all the retry attempts and is now going to move on (and from telemetry purposes I want a separate event in case we change the number of retries later - so that we can easily pick out the number of times that a user hit the worst-case scenario where the retries all failed).

So what I'd suggest is to : 

1. Remove the when from this catch
2. Have an if in here that changes what is logged/tracked based on whether it's the last attempt or not
3. Remove the outer catch 
4. You should be able to just remove the finally as well and have it run after the loop - since we'll be catching and handling all the exceptions